tls/s2n_shutdown.c

```diff
@@ -41,81 +36,97 @@ static bool s2n_error_alert_received(struct s2n_connection *conn)
     return true;
 }
 
-static bool s2n_error_alert_sent(struct s2n_connection *conn)
+static bool s2n_shutdown_expect_close_notify(struct s2n_connection *conn)
 {
-    /* Sending an alert always sets conn->write_closing: see s2n_flush() */
-    if (!conn->write_closing) {
+    /* No close_notify expected if we already received an error instead */
+    if (s2n_error_alert_received(conn)) {
+        return false;
+    }
+
+    /* No close_notify expected if we sent an error instead of a close_notify */
+    if (conn->writer_alert_out || conn->reader_alert_out) {
         return false;
     }
-    /* Verify that the alert sent wasn't just a close_notify */
-    if (conn->close_notify_queued) {
+
+    /* The purpose of the peer responding to our close_notify
+     * with its own close_notify is to prevent application data truncation.
+     * However, application data is not a concern during the handshake.
+     *
+     * Additionally, decrypting alerts sent during the handshake can be error prone
+     * due to different encryption keys and may lead to unnecessary error reporting
+     * and unnecessary blinding.
+     */
+    if (!s2n_handshake_is_complete(conn)) {
         return false;
     }
+
+    /* QUIC does not use TLS alerts */
+    if (conn->quic_enabled) {
```

Comment:
Should we use `s2n_alerts_supported` here?

Reply:
s2n_alerts_supported is currently static, and I'm not sure it's worth changing that for this check. Do you think it's worth it?